### PR TITLE
feat(core): introduce nx import

### PR DESCRIPTION
**Pull request by FrozenPandaz**
## Current Behavior

Importing other projects/ repositories into an Nx workspace is a natural part of the Nx adoption story. However, there is no easy built in way of handling this while maintaining `git` history.

## Expected Behavior

`nx import` is a new command which allows teams to merge code from other repositories into a Nx workspace.

https://asciinema.org/a/oQiA9qOvA2z85AQvVJ5QRVTp1

## Related Issue(s)

Fixes #

### CODE REVIEW

File: packages/nx/src/command-line/import/import.ts [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+  const tempImportDirectory = join(tmpdir, 'nx-import');
+
+  if (!sourceRemoteUrl) {
+    sourceRemoteUrl = (
+      await prompt<{ sourceRemoteUrl: string }>([
+        {
+          type: 'input',
+          name: 'sourceRemoteUrl',
+          message:
+            'What is the Remote URL of the repository you want to import?',
```
Comment (jaysoo) [inline packages/nx/src/command-line/import/import.ts:80]: We can rephrase this to also include local directories.

Comment (jaysoo) [reply packages/nx/src/command-line/import/import.ts:80]: Should test and verify the path passed if it's local

File: packages/nx/src/command-line/import/import.ts [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+           */
+          limit: process.stdout.rows - 3,
+          required: true,
+        } as any,
+      ])
+    ).ref;
+  }
+
+  if (!source) {
+    source = (
```
Comment (jaysoo) [inline packages/nx/src/command-line/import/import.ts:122]: Validate this exists.

Q: Can we valdiate the project?

File: packages/nx/src/command-line/import/utils/prepare-source-repo.ts [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+  source: string,
+  relativeDestination: string,
+  tempImportBranch: string,
+  sourceRemoteUrl: string
+) {
+  const spinner = createSpinner().start(
+    `Fetching ${ref} from ${sourceRemoteUrl}`
+  );
+  await gitClient.addFetchRemote('origin', ref);
+  await gitClient.fetch('origin', { depth: 1, ref });
```
Comment (jaysoo) [inline packages/nx/src/command-line/import/utils/prepare-source-repo.ts:23]: Check if all the history if `depth: 1`

File: packages/nx/src/command-line/import/utils/prepare-source-repo.ts [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+        continue;
+      }
+      spinner.start(
+        `Moving files and git history to ${destinationInSource}: ${file}`
+      );
+
+      const newPath = join(destinationInSource, file);
+
+      await mkdirAsync(dirname(newPath), { recursive: true });
+      try {
```
Comment (jaysoo) [inline packages/nx/src/command-line/import/utils/prepare-source-repo.ts:61]: If we're moving to a gitignored directory, what will happen?

File: e2e/nx/src/import.test.ts [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+    () =>
+      (proj = newProject({
+        packages: ['@nx/js'],
+        unsetProjectNameAndRootFormat: false,
+      }))
+  );
+  afterAll(() => cleanupProject());
+
+  it('should be able to import a vite app', () => {
+    const remote = 'https://github.com/nrwl/nx.git';
```
Comment (jaysoo) [inline e2e/nx/src/import.test.ts:23]: Maybe use a local remote so it's not necessary to have a reference on remote?

File: packages/nx/src/command-line/import/import.ts (hunk 2 above)
Comment (FrozenPandaz) [reply packages/nx/src/command-line/import/import.ts:126]: Validated

File: packages/nx/src/command-line/import/utils/prepare-source-repo.ts (hunk 3 above)
Comment (FrozenPandaz) [reply packages/nx/src/command-line/import/utils/prepare-source-repo.ts]: Yup, that messed it up. I made it the full history... 

If it's a local path... it kind of adopts the depth of the local path... so if people have a shallow clone, then they might be missing some commits :thinking: 

But if it's a remote path, we get the full history, however long it is

File: e2e/nx/src/import.test.ts (hunk 5 above)
Comment (FrozenPandaz) [reply e2e/nx/src/import.test.ts]: Done

**Review by jaysoo — APPROVED** [review]

File: packages/nx/src/command-line/import/import.ts [hunk 6]
```diff
@@ -0,0 +1,270 @@
...
+  const sourceRepoPath = join(tempImportDirectory, 'repo');
+  const spinner = createSpinner(
+    `Cloning ${sourceRemoteUrl} into a temporary directory: ${sourceRepoPath}`
+  ).start();
+  try {
+    await rm(tempImportDirectory, { recursive: true });
+  } catch {}
+  await mkdir(tempImportDirectory, { recursive: true });
+
+  const sourceGitClient = await cloneFromUpstream(
```
Comment (jaysoo) [inline packages/nx/src/command-line/import/import.ts:100]: If this fails, we may want to ask user to check their URL / directory path. If there is a typo, or what they are importing isn't a git repo.
